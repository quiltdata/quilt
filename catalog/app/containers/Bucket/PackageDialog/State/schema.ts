import type { ErrorObject } from 'ajv'
import * as React from 'react'
import * as R from 'ramda'

import * as AWS from 'utils/AWS'
import {
  JsonSchema,
  makeSchemaDefaultsSetter,
  makeSchemaValidator,
} from 'utils/JSONSchema'
import * as Request from 'utils/useRequest'
import * as Types from 'utils/types'
import * as workflows from 'utils/workflows'

import { metadataSchema, objectSchema } from '../../requests'

export type SchemaStatus =
  | { _tag: 'idle' }
  | { _tag: 'loading' }
  | { _tag: 'error'; error: Error }
  | { _tag: 'ready'; schema?: JsonSchema }

export const Idle = { _tag: 'idle' as const }
export const Loading = { _tag: 'loading' as const }
export const Err = (error: Error) => ({ _tag: 'error' as const, error })
export const Ready = (schema?: JsonSchema) => ({ _tag: 'ready' as const, schema })

export function mkMetaValidator(schema?: JsonSchema) {
  const schemaValidator = makeSchemaValidator(schema)
  return function validateMeta(value: Types.Json): (ErrorObject | Error)[] | undefined {
    const jsonObjectErr = value && !R.is(Object, value)
    if (jsonObjectErr) {
      return [new Error('Metadata must be a valid JSON object')]
    }

    const setDefaults = makeSchemaDefaultsSetter(schema)
    const errors = schemaValidator(setDefaults(value))
    if (errors.length) return errors
  }
}

export function useMetadataSchema(workflow?: workflows.Workflow): SchemaStatus {
  const s3 = AWS.S3.use()
  const schemaUrl = workflow?.schema?.url
  const req = React.useCallback(() => metadataSchema({ s3, schemaUrl }), [schemaUrl, s3])
  const result = Request.use(req, !!schemaUrl)

  if (!schemaUrl) return Ready()

  if (result === Request.Idle) return Idle
  if (result === Request.Loading) return Loading
  if (result instanceof Error) return Err(result)

  return Ready(result)
}

export function useEntriesSchema(workflow?: workflows.Workflow): SchemaStatus {
  const s3 = AWS.S3.use()
  const schemaUrl = workflow?.entriesSchema || ''
  const req = React.useCallback(() => objectSchema({ s3, schemaUrl }), [schemaUrl, s3])
  const result = Request.use(req, !!schemaUrl)

  if (!schemaUrl) return Ready()

  if (result === Request.Idle) return Idle
  if (result === Request.Loading) return Loading
  if (result instanceof Error) return Err(result)

  return Ready(result)
}
