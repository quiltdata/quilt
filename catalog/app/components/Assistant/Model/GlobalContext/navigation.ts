import * as Eff from 'effect'
import * as React from 'react'
import * as RR from 'react-router-dom'
import { Schema as S } from '@effect/schema'

import bucketRoutes from 'containers/Bucket/Routes'
import search from 'containers/Search/Route'
import * as ROUTES from 'constants/routes'
import * as Log from 'utils/Logging'
import * as Nav from 'utils/Navigation'
import * as XML from 'utils/XML'

import * as Content from '../Content'
import * as Context from '../Context'
import * as Tool from '../Tool'

const MODULE = 'GlobalContext/navigation'

// the routes are in the order of matching
// TODO: specify/describe all the *relevant* routes
const routeList = [
  Nav.makeRoute({
    name: 'home',
    path: ROUTES.home.path,
    exact: true,
    description: 'Home page',
    // searchParams: S.Struct({
    //   // XXX: passing this param doesn't actually work bc of how it's implemented in
    //   //      website/pages/Landing/Buckets/Buckets.js
    //   q: SearchParamLastOpt.annotations({
    //     title: 'bucket filter query',
    //     description: 'filter buckets in the bucket grid',
    //   }),
    // }),
  }),
  Nav.makeRoute({
    name: 'install',
    path: ROUTES.install.path,
    description: 'Installation page',
  }),
  search,
  Nav.makeRoute({
    name: 'activate',
    path: ROUTES.activate.path,
    description: 'TBD',
  }),
  //         <Route path={paths.signIn} exact>
  //           <AuthSignIn />
  //         </Route>
  //         <Route path="/login" exact>
  //           <RedirectTo path={urls.signIn()} />
  //         </Route>
  //
  //         <Route path={paths.signOut} exact>
  //           <AuthSignOut />
  //         </Route>
  //
  //         {(cfg.passwordAuth === true || cfg.ssoAuth === true) && (
  //           <Route path={paths.signUp} exact>
  //             <AuthSignUp />
  //           </Route>
  //         )}
  //         {!!cfg.passwordAuth && (
  //           <Route path={paths.passReset} exact>
  //             <AuthPassReset />
  //           </Route>
  //         )}
  //         {!!cfg.passwordAuth && (
  //           <Route path={paths.passChange} exact>
  //             <AuthPassChange />
  //           </Route>
  //         )}
  //
  //         <Route path={paths.code} exact>
  //           <AuthCode />
  //         </Route>
  //
  //         <Route path={paths.activationError} exact>
  //           <AuthActivationError />
  //         </Route>
  //
  //         {cfg.mode === 'OPEN' && (
  //           // XXX: show profile in all modes?
  //           <Route path={paths.profile} exact>
  //             <OpenProfile />
  //           </Route>
  //         )}
  //
  //         <Route path={paths.admin}>
  //           <Admin />
  //         </Route>
  //
  //         <Route path={paths.uriResolver}>
  //           <UriResolver />
  //         </Route>
  //
  ...bucketRoutes,
] as const

type KnownRoute = (typeof routeList)[number]
type KnownRouteMap = {
  [K in KnownRoute['name']]: Extract<KnownRoute, { name: K }>
}
export const routes = Object.fromEntries(
  routeList.map((r) => [r.name, r]),
) as KnownRouteMap

export const NavigableRouteSchema = S.Union(
  ...routeList.map((r) => r.navigableRouteSchema),
)

type NavigableRoute = typeof NavigableRouteSchema.Type

export const locationFromRoute = (route: NavigableRoute) =>
  // @ts-expect-error
  S.encode(routes[route.name].paramsSchema)(route.params)

type History = ReturnType<typeof RR.useHistory>

const WAIT_TIMEOUT = Eff.Duration.seconds(30)
const NAV_LAG = Eff.Duration.seconds(1)

const navigate = (
  route: NavigableRoute,
  history: History,
  markers: Eff.SubscriptionRef.SubscriptionRef<Record<string, boolean>>,
) =>
  Log.scoped({
    name: `${MODULE}.navigate`,
    enter: [`to: ${route.name}`, Log.br, 'params:', route.params],
  })(
    Eff.pipe(
      locationFromRoute(route),
      Eff.Effect.tap((loc) => Eff.Effect.log(`Navigating to location:`, Log.br, loc)),
      Eff.Effect.andThen((loc) => Eff.Effect.sync(() => history.push(loc))),
      Eff.Effect.andThen(() =>
        Eff.Effect.gen(function* () {
          const { waitForMarkers } = routes[route.name]
          if (!waitForMarkers.length) return
          yield* Eff.Effect.log(`Waiting for markers: ${waitForMarkers.join(', ')}`)
          yield* Eff.Effect.sleep(NAV_LAG)
          yield* Eff.pipe(
            markers.changes,
            Eff.Stream.timeoutFail(() => ({ _tag: 'timeout' as const }), WAIT_TIMEOUT),
            Eff.Stream.runForEachWhile((currentMarkers) =>
              Eff.Effect.succeed(!waitForMarkers.every((k) => currentMarkers[k])),
            ),
            Eff.Effect.andThen(() => Eff.Effect.log('Markers found')),
            Eff.Effect.catchTag('timeout', () =>
              Eff.Effect.log(
                `Timed out after ${Eff.Duration.format(
                  WAIT_TIMEOUT,
                )} while waiting for markers`,
              ),
            ),
          )
        }),
      ),
    ),
  )

export interface Match {
  descriptor: KnownRoute
  decoded: NavigableRoute | null
}

export const matchLocation = (loc: typeof Nav.Location.Type): Match | null =>
  Eff.pipe(
    Eff.Array.findFirst(routeList, (route) =>
      RR.matchPath(loc.pathname, {
        path: route.path,
        exact: route.exact,
        strict: route.strict,
      })
        ? Eff.Option.some(route)
        : Eff.Option.none(),
    ),
    Eff.Option.map((descriptor) => ({
      descriptor,
      decoded: Eff.pipe(
        loc,
        // @ts-expect-error
        S.decodeOption(descriptor.paramsSchema),
        Eff.Option.map((params) => ({ name: descriptor.name, params }) as NavigableRoute),
        Eff.Option.getOrNull,
      ),
    })),
    Eff.Option.getOrNull,
  )

export function useRouteContext() {
  const loc = RR.useLocation()

  const match = React.useMemo(
    () => matchLocation({ pathname: loc.pathname, search: loc.search, hash: '' }),
    [loc.pathname, loc.search],
  )

  const description = React.useMemo(() => {
    if (!match) return ''
    const params = match.decoded?.params
      ? XML.tag('parameters', {}, JSON.stringify(match.decoded.params, null, 2))
      : null
    return XML.tag(
      'route-info',
      {},
      `Name: "${match.descriptor.name}"`,
      XML.tag('description', {}, match.descriptor.description),
      params,
    )
  }, [match])

  const msg = React.useMemo(
    () =>
      XML.tag(
        'viewport',
        {},
        XML.tag('current-location', {}, JSON.stringify(loc, null, 2)),
        description,
        'Refer to "navigate" tool schema for navigable routes and their parameters.',
      ).toString(),
    [description, loc],
  )

  return msg
}

export const NavigateSchema = S.Struct({
  route: NavigableRouteSchema,
}).annotations({
  title: 'navigate the catalog',
  description: 'navigate to a provided route',
})

export function useNavigate() {
  const history = RR.useHistory()
  const markers = Context.useMarkersRef()

  return Tool.useMakeTool(
    NavigateSchema,
    ({ route }) =>
      Eff.pipe(
        navigate(route, history, markers),
        Eff.Effect.match({
          onSuccess: () =>
            Tool.succeed(Content.text(`Navigating to the '${route.name}' route.`)),
          onFailure: (e) =>
            Tool.fail(
              Content.text(`Failed to navigate to the '${route.name}' route: ${e}`),
            ),
        }),
        Eff.Effect.map(Eff.Option.some),
      ),
    [history, markers],
  )
}
