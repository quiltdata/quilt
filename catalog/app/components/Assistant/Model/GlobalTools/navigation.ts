import * as Eff from 'effect'
import * as RR from 'react-router-dom'
import { Schema as S } from '@effect/schema'

import * as ROUTES from 'constants/routes'
import * as Log from 'utils/Logging'
import * as Nav from 'utils/Navigation'

import * as Content from '../Content'
import * as Tool from '../Tool'

const MODULE = 'GlobalTools/navigation'

const home = Nav.makeRoute({
  name: 'home',
  path: ROUTES.home.path,
  description: 'Home page',
  // searchParams: S.Struct({
  //   // XXX: passing this param doesn't actually work bc of how it's implemented in
  //   //      website/pages/Landing/Buckets/Buckets.js
  //   q: SearchParamLastOpt.annotations({
  //     title: 'bucket filter query',
  //     description: 'filter buckets in the bucket grid',
  //   }),
  // }),
})

const search = Nav.makeRoute({
  name: 'search' as const,
  path: ROUTES.home.path,
  description: 'Search page',
  // TODO: add search params
  // pathParams: emptyPathParams,
  // searchParams: emptySearchParams,
})

const routeList = [home, search] as const
type KnownRoute = (typeof routeList)[number]
type KnownRouteMap = {
  [K in KnownRoute['name']]: Extract<KnownRoute, { name: K }>
}
const routes = Object.fromEntries(routeList.map((r) => [r.name, r])) as KnownRouteMap

const NavigableRouteSchema = S.Union(...routeList.map((r) => r.navigableRouteSchema))

type NavigableRoute = typeof NavigableRouteSchema.Type

const NavigateSchema = S.Struct({
  route: NavigableRouteSchema,
}).annotations({
  title: 'navigate the catalog',
  description: 'navigate to a provided route',
})

type History = ReturnType<typeof RR.useHistory>

const navigate = (route: NavigableRoute, history: History) =>
  Log.scoped({
    name: `${MODULE}.navigate`,
    enter: [`to: ${route.name}`, Log.br, 'params:', route.params],
  })(
    Eff.pipe(
      route.params,
      S.encode(routes[route.name].paramsSchema),
      Eff.Effect.andThen((loc) => Eff.Effect.sync(() => history.push(loc))),
      Eff.Effect.match({
        onSuccess: () =>
          Tool.succeed(Content.text(`Navigating to the '${route.name}' route.`)),
        onFailure: (e) =>
          Tool.fail(
            Content.text(`Failed to navigate to the '${route.name}' route: ${e}`),
          ),
      }),
      Eff.Effect.map(Eff.Option.some),
    ),
  )

export function useNavigate() {
  const history = RR.useHistory()
  return Tool.useMakeTool(NavigateSchema, ({ route }) => navigate(route, history), [
    history,
  ])
}
