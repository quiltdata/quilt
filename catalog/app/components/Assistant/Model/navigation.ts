import * as Eff from 'effect'
import * as RR from 'react-router-dom'
import { Schema as S } from '@effect/schema'

import search from 'containers/Search/Route'
import * as ROUTES from 'constants/routes'
import * as Log from 'utils/Logging'
import * as Nav from 'utils/Navigation'

const MODULE = 'Assistant/Model/navigation'

const home = Nav.makeRoute({
  name: 'home',
  path: ROUTES.home.path,
  description: 'Home page',
  // searchParams: S.Struct({
  //   // XXX: passing this param doesn't actually work bc of how it's implemented in
  //   //      website/pages/Landing/Buckets/Buckets.js
  //   q: SearchParamLastOpt.annotations({
  //     title: 'bucket filter query',
  //     description: 'filter buckets in the bucket grid',
  //   }),
  // }),
})

const routeList = [home, search] as const
type KnownRoute = (typeof routeList)[number]
type KnownRouteMap = {
  [K in KnownRoute['name']]: Extract<KnownRoute, { name: K }>
}
const routes = Object.fromEntries(routeList.map((r) => [r.name, r])) as KnownRouteMap

export const NavigableRouteSchema = S.Union(
  ...routeList.map((r) => r.navigableRouteSchema),
)

export type NavigableRoute = typeof NavigableRouteSchema.Type

export type History = ReturnType<typeof RR.useHistory>

export const navigate = (route: NavigableRoute, history: History) =>
  Log.scoped({
    name: `${MODULE}.navigate`,
    enter: [`to: ${route.name}`, Log.br, 'params:', route.params],
  })(
    Eff.pipe(
      route.params,
      // @ts-expect-error
      S.encode(routes[route.name].paramsSchema),
      Eff.Effect.tap((loc) => Eff.Effect.log(`Navigating to location:`, Log.br, loc)),
      Eff.Effect.andThen((loc) => Eff.Effect.sync(() => history.push(loc))),
    ),
  )
