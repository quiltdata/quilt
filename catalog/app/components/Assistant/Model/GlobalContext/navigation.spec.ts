import * as Eff from 'effect'
import { JSONSchema, Schema } from '@effect/schema'

import * as nav from './navigation'

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

describe('components/Assistant/Model/GlobalTools/navigation', () => {
  describe('NavigateSchema', () => {
    describe('produced JSON Schema', () => {
      it('should match the snapshot', () => {
        const jsonSchema = JSONSchema.make(nav.NavigateSchema)
        expect(jsonSchema).toMatchSnapshot()
      })
    })
  })
  describe('routes', () => {
    const TEST_CASES = [
      {
        route: {
          name: 'search',
          params: {
            searchString: '',
            buckets: [],
            order: 'NEWEST',
            params: {
              resultType: 'p',
              filter: [],
              latestOnly: true,
              userMetaFilters: [
                {
                  path: '/author',
                  predicate: {
                    type: 'KeywordEnum',
                    value: {
                      terms: ['Aneesh', 'Maksim'],
                    },
                  },
                },
              ],
            },
          },
        },
        loc: {
          pathname: '/search',
          search: 'o=NEWEST&meta.e%2Fauthor=%22Aneesh%22%2C%22Maksim%22',
          hash: '',
        },
      },
      {
        route: {
          name: 'bucket.overview',
          params: {
            bucket: 'test-bucket',
          },
        },
        loc: {
          pathname: '/b/test-bucket',
          search: '',
          hash: '',
        },
      },
      {
        route: {
          name: 'bucket.prefix',
          params: {
            bucket: 'quilt-example',
            path: 'data/random-data-benchmark/100kb/',
          },
        },
        loc: {
          pathname: '/b/quilt-example/tree/data/random-data-benchmark/100kb/',
          search: '',
          hash: '',
        },
      },
    ]

    const encode = Eff.flow(
      Schema.decodeUnknown(nav.NavigableRouteSchema),
      Eff.Effect.andThen(nav.locationFromRoute),
      Eff.Effect.runPromise,
    )

    TEST_CASES.forEach((tc, i) => {
      describe(`${i + 1}: ${tc.route.name}`, () => {
        it('should encode', async () => {
          expect(await encode(tc.route)).toEqual(tc.loc)
        })
        it('should decode', async () => {
          expect(nav.matchLocation(tc.loc)?.decoded).toEqual(tc.route)
        })
      })
    })
  })
})
