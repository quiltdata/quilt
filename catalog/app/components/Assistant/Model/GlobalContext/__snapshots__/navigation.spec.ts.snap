// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Assistant/Model/GlobalTools/navigation NavigateSchema produced JSON Schema should match the snapshot 1`] = `
{
  "$defs": {
    "BooleanPredicateState": {
      "additionalProperties": false,
      "properties": {
        "false": {
          "type": "boolean",
        },
        "true": {
          "type": "boolean",
        },
      },
      "required": [],
      "title": "Boolean predicate state",
      "type": "object",
    },
    "BooleanTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "Boolean",
          ],
        },
        "value": {
          "$ref": "#/$defs/BooleanPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "DatetimePredicateState": {
      "additionalProperties": false,
      "properties": {
        "gte": {
          "description": "ISO date string",
          "format": "date-time",
          "type": "string",
        },
        "lte": {
          "description": "ISO date string",
          "format": "date-time",
          "type": "string",
        },
      },
      "required": [],
      "title": "Datetime predicate state",
      "type": "object",
    },
    "DatetimeTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "Datetime",
          ],
        },
        "value": {
          "$ref": "#/$defs/DatetimePredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "KeywordEnumPredicateState": {
      "additionalProperties": false,
      "properties": {
        "terms": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
      },
      "required": [],
      "title": "KeywordEnum predicate state",
      "type": "object",
    },
    "KeywordEnumTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "KeywordEnum",
          ],
        },
        "value": {
          "$ref": "#/$defs/KeywordEnumPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "KeywordWildcardPredicateState": {
      "additionalProperties": false,
      "properties": {
        "wildcard": {
          "type": "string",
        },
      },
      "required": [],
      "title": "KeywordWildcard predicate state",
      "type": "object",
    },
    "KeywordWildcardTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "KeywordWildcard",
          ],
        },
        "value": {
          "$ref": "#/$defs/KeywordWildcardPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "NumberPredicateState": {
      "additionalProperties": false,
      "properties": {
        "gte": {
          "type": "number",
        },
        "lte": {
          "type": "number",
        },
      },
      "required": [],
      "title": "Number predicate state",
      "type": "object",
    },
    "NumberTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "Number",
          ],
        },
        "value": {
          "$ref": "#/$defs/NumberPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
    "TextPredicateState": {
      "additionalProperties": false,
      "properties": {
        "queryString": {
          "type": "string",
        },
      },
      "required": [],
      "title": "Text predicate state",
      "type": "object",
    },
    "TextTaggedPredicate": {
      "additionalProperties": false,
      "properties": {
        "type": {
          "enum": [
            "Text",
          ],
        },
        "value": {
          "$ref": "#/$defs/TextPredicateState",
        },
      },
      "required": [
        "type",
        "value",
      ],
      "type": "object",
    },
  },
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "description": "navigate to a provided route",
  "properties": {
    "route": {
      "anyOf": [
        {
          "additionalProperties": false,
          "description": "Home page",
          "properties": {
            "name": {
              "enum": [
                "home",
              ],
            },
            "params": {
              "$id": "/schemas/{}",
              "anyOf": [
                {
                  "type": "object",
                },
                {
                  "type": "array",
                },
              ],
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "Installation page",
          "properties": {
            "name": {
              "enum": [
                "install",
              ],
            },
            "params": {
              "$id": "/schemas/{}",
              "anyOf": [
                {
                  "type": "object",
                },
                {
                  "type": "array",
                },
              ],
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "Search page",
          "properties": {
            "name": {
              "enum": [
                "search",
              ],
            },
            "params": {
              "additionalProperties": false,
              "properties": {
                "buckets": {
                  "description": "A list of buckets to search in (keep empty to search in all buckets)",
                  "items": {
                    "type": "string",
                  },
                  "title": "Search buckets",
                  "type": "array",
                },
                "order": {
                  "$comment": "/schemas/enums",
                  "anyOf": [
                    {
                      "enum": [
                        "BEST_MATCH",
                      ],
                      "title": "BEST_MATCH",
                    },
                    {
                      "enum": [
                        "NEWEST",
                      ],
                      "title": "NEWEST",
                    },
                    {
                      "enum": [
                        "OLDEST",
                      ],
                      "title": "OLDEST",
                    },
                  ],
                  "description": "Order of search results",
                  "title": "Search result order",
                },
                "params": {
                  "anyOf": [
                    {
                      "additionalProperties": false,
                      "properties": {
                        "filter": {
                          "description": "Filter results by system metadata",
                          "items": {
                            "anyOf": [
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "modified",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/DatetimePredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Last modified date",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "size",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/NumberPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "File size in bytes",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "name",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordWildcardPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Package name (aka namespace or handle)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "hash",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordWildcardPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Package revision hash",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "entries",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/NumberPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Number of package entries",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "comment",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/TextPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Package revision comment (aka commit message)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "workflow",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordEnumPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Package workflow",
                                "type": "object",
                              },
                            ],
                          },
                          "title": "Result filters (system metadata)",
                          "type": "array",
                        },
                        "latestOnly": {
                          "description": "Search only latest revisions",
                          "title": "Latest only",
                          "type": "boolean",
                        },
                        "resultType": {
                          "enum": [
                            "p",
                          ],
                          "title": "Result type: Quilt Package",
                        },
                        "userMetaFilters": {
                          "description": "Filter results by user metadata",
                          "items": {
                            "additionalProperties": false,
                            "properties": {
                              "path": {
                                "type": "string",
                              },
                              "predicate": {
                                "anyOf": [
                                  {
                                    "$ref": "#/$defs/BooleanTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/DatetimeTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/KeywordEnumTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/KeywordWildcardTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/NumberTaggedPredicate",
                                  },
                                  {
                                    "$ref": "#/$defs/TextTaggedPredicate",
                                  },
                                ],
                              },
                            },
                            "required": [
                              "path",
                              "predicate",
                            ],
                            "type": "object",
                          },
                          "title": "Result filters (user metadata)",
                          "type": "array",
                        },
                      },
                      "required": [
                        "resultType",
                      ],
                      "title": "Package-specific search parameters",
                      "type": "object",
                    },
                    {
                      "additionalProperties": false,
                      "properties": {
                        "filter": {
                          "items": {
                            "anyOf": [
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "modified",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/DatetimePredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Last modified date",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "size",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/NumberPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "File size in bytes",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "ext",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordEnumPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "File extensions (with a leading dot)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "key",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/KeywordWildcardPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "File name (aka S3 Object Key)",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "content",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/TextPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Indexed text contents",
                                "type": "object",
                              },
                              {
                                "additionalProperties": false,
                                "properties": {
                                  "key": {
                                    "enum": [
                                      "deleted",
                                    ],
                                  },
                                  "predicate": {
                                    "$ref": "#/$defs/BooleanPredicateState",
                                  },
                                },
                                "required": [
                                  "key",
                                  "predicate",
                                ],
                                "title": "Whether a file is a delete marker",
                                "type": "object",
                              },
                            ],
                          },
                          "title": "Result filters",
                          "type": "array",
                        },
                        "resultType": {
                          "enum": [
                            "o",
                          ],
                          "title": "Result type: S3 Object",
                        },
                      },
                      "required": [
                        "resultType",
                      ],
                      "title": "Object-specific search parameters",
                      "type": "object",
                    },
                  ],
                  "title": "Result type-specific parameters",
                },
                "searchString": {
                  "description": "A String to search for. ElasticSearch syntax supported. For packages, searches in package name, comment (commit message), and metadata. For objects, searches in object key and indexed content.",
                  "title": "Search string",
                  "type": "string",
                },
              },
              "required": [
                "params",
              ],
              "type": "object",
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "TBD",
          "properties": {
            "name": {
              "enum": [
                "activate",
              ],
            },
            "params": {
              "$id": "/schemas/{}",
              "anyOf": [
                {
                  "type": "object",
                },
                {
                  "type": "array",
                },
              ],
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "S3 Object (aka File) page",
          "properties": {
            "name": {
              "enum": [
                "bucket.object",
              ],
            },
            "params": {
              "additionalProperties": false,
              "properties": {
                "bucket": {
                  "type": "string",
                },
                "mode": {
                  "description": "Contents preview mode",
                  "title": "Viewing Mode",
                  "type": "string",
                },
                "path": {
                  "description": "S3 Object Key aka File Path",
                  "title": "Path",
                  "type": "string",
                },
                "version": {
                  "description": "S3 Object Version ID (omit for latest version)",
                  "title": "Version ID",
                  "type": "string",
                },
              },
              "required": [
                "bucket",
                "path",
              ],
              "type": "object",
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "S3 Prefix (aka Directory) page",
          "properties": {
            "name": {
              "enum": [
                "bucket.prefix",
              ],
            },
            "params": {
              "additionalProperties": false,
              "properties": {
                "bucket": {
                  "type": "string",
                },
                "path": {
                  "description": "S3 Prefix aka Directory Path",
                  "title": "Path",
                  "type": "string",
                },
              },
              "required": [
                "bucket",
                "path",
              ],
              "type": "object",
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
        {
          "additionalProperties": false,
          "description": "Bucket overview page",
          "properties": {
            "name": {
              "enum": [
                "bucket.overview",
              ],
            },
            "params": {
              "additionalProperties": false,
              "properties": {
                "bucket": {
                  "type": "string",
                },
              },
              "required": [
                "bucket",
              ],
              "type": "object",
            },
          },
          "required": [
            "name",
            "params",
          ],
          "type": "object",
        },
      ],
    },
  },
  "required": [
    "route",
  ],
  "title": "navigate the catalog",
  "type": "object",
}
`;
